### ISSUE TITLE
no-array-index-key allows strings with the id in them

### ISSUE THREAD

**Issue by buesing** [yannickcr/eslint-plugin-react#2081]
This rule has led to a lot of devs in our company using something like
```
{items.map((item, i) => <Item key={`key_${i}`} />)}
```
which satisfies the rule, but suffers from the same problems as just using the index as a key. Not sure if this is feasible but ideally the rule should recognize if the key is derived only from the index, and in this case throw a warning.

**Comment by ljharb** [comment]
I'm not really sure how it's feasible. To be sure, a hardcoded string key where the only variable part is the index is the same as using only the index, and that should be avoided.

**Comment by buesing** [comment]
@ljharb my bad, the rule already recognizes this case. All good there. Something else I noticed: I often use a class method to render items from an array, like this:
```
  renderParagraph = (paragraph, i) => {
    return (
      <p key={i}>
        {paragraph}
      </p>
    );
  }

  render() {
    return (
      <article>
        {paragraphs.map(this.renderParagraph)}
      </article>
    );
  }
```
This currently doesn't throw an error.

**Comment by ljharb** [comment]
What happens if you use an actual method, instead of an arrow function in a class field (which I’d recommend never doing)?

**Comment by buesing** [comment]
You mean like this:
```
  renderParagraph(paragraph, i) {
    return (
      <div key={i}>
        {paragraph}
      </div>
    );
  }

  render() {
    return (
      <article>
        {data.map(this.renderParagraph)}
      </article>
    );
  }
```
Same thing, the rule doesn't complain.
If you don't mind, could you explain why you recommend never using an arrow function in a class field?

**Comment by ljharb** [comment]
Let’s file a new issue for both of those not complaining; I’m going to close this one.

Using an arrow function in a class field prevents the problem of sending different === references to the function in every render (as opposed to inline functions); but it makes testing harder (because you can’t mock out the implementation before rendering) and prevents optimizations by creating a new function per instance, instead of sharing the “meat” on the prototype and only creating a .bind per instance.
